Update BeamSpot scripts for FTVs

francescobrivio opened the pull request:
#### PR description:
In order to ease the BeamSpot workflow usage by the TkAl experts for the Full Track Validations (in order not to rely only on the BeamSpot experts), this PR updates two scripts:
- `RecoVertex/BeamSpotProducer/scripts/createPayload.py`
   - The `option.merged` is used to handle the cases in which the input txt file contains multiple BeamSpot fits. This option was never really used so far (AFAIK) and it contained a few bugs, which are now fixed.
- The `BeamSpotOnlineFromOfflineConverter.cc` plugin (and its python config) has been updated to be able to also handle multi-IOV input tags via the `ESWatcher` mechanism (thanks @mmusich for the tip!)

**Important Note:**
Both these scripts are **NOT** used in production, but only ran manually by experts.

#### PR validation:
Few tests run:
 - code compiles
 - ran `scram b runtests` successfully
 - tested privately with:
   -  ```
      python3 createPayload.py -I lumibase -t [MY_TAG] -d [INPUT_TXT_FILE] -O [OUTPUT_DIR] --merged
      ```
   -  ```
      cmsRun BeamSpotOnlineFromOfflineConverter_cfg.py unitTest=True
      ```
   - ```
      cmsRun BeamSpotOnlineFromOfflineConverter_cfg.py \
        inputTag=[INPUT_TAG] \
        inputFile=[INPUT_DB_FILE] \
        outputTag=[OUTPUT_TAG] \
        startRun=373710 \
        startLumi=179
      ```

#### Backport:
Not a backport, no backport needed

------
FYI @lguzzi @dzuolo @mmusich @TomasKello @henriettepetersen

francescobrivio committed:
fix script createPayload.py for BeamSpot

francescobrivio commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42909/37051

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @francescobrivio  for master.

It involves the following packages:

- CondTools/BeamSpot (**alca, db**)
- RecoVertex/BeamSpotProducer (**alca, reconstruction**)


@francescobrivio, @jfernan2, @saumyaphor4252, @perrotta, @consuegs, @mandrenguyen can you please review it and eventually sign? Thanks.
@dgulhan, @mtosi, @VourMa, @missirol, @mmusich, @tocheng, @GiacomoSguazzoni, @rovere, @VinInn, @JanFSchulte this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4efcb7/34973/summary.html
**COMMIT**: a453b6e8d20e8666b707c2df0e586f9290065e4c
**CMSSW**: CMSSW_13_3_X_2023-09-28-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42909/34973/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 1 lines from the logs
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3358320
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3358298
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-09-28-1100+4efcb7/59073/triggerResults): no differences found

francescobrivio commented:
+db

mmusich commented on CondTools/BeamSpot/plugins/BeamSpotOnlineFromOfflineConverter.cc:
```diff
@@ -115,56 +117,61 @@ cond::Time_t BeamSpotOnlineFromOfflineConverter::pack(uint32_t fIOVStartRun, uin
-      } else
-        poolDbService->appendOneIOV<BeamSpotOnlineObjects>(abeam, poolDbService->currentTime(), fLabel);
+  if (bsWatcher_.check(iSetup)) {
+    const std::string fLabel = (fIsHLT_) ? "BeamSpotOnlineHLTObjectsRcd" : "BeamSpotOnlineLegacyObjectsRcd";
```
I know this was already like that before, but this could go in the class constructor, right?

francescobrivio replied:
sure will do in the next commit!

francescobrivio replied:
addressed in last force-push commit: https://github.com/cms-sw/cmssw/pull/42909/commits/064118e709f10e968d446bb3d53662762a7e8606

mmusich reviewed (commented):
would you mind adding a unit test for the new mode 

https://github.com/cms-sw/cmssw/blob/10937c351ebec5d7e1be3250fc62bb585f7434c1/CondTools/BeamSpot/test/testReadWriteBeamSpotsFromDB.sh#L44

(and fix the bug in the output string I left there ;)

francescobrivio commented:
> would you mind adding a unit test for the new mode

At the moment the unitTest is reading the `auto:phase2_realistic` MC GT which of course has a a single IOV.
The easiest solution to update the unitTest would be to read  a data GT, but at the cost of reading (and writing out) a lot of IOVs...would that be acceptable? Or do you have any other suggestion on how to implement it?

mmusich commented:
>  but at the cost of reading (and writing out) a lot of IOVs...would that be acceptable?

can't you limit the amount read / writes just by tuning the input parameters of the script (e.g.`startRun` and `maxEvents`?)

francescobrivio commented:
Since I'll be offline the next 3 days I have pushed two of the changes requested:
 - moved `fLabel` initialization to the constructor
 - fixed bug in the unitTest printout message

Once we converge I on how to improve the unitTest I'll make a second force-push (or add it in a next PR).

francescobrivio commented:
> > but at the cost of reading (and writing out) a lot of IOVs...would that be acceptable?
> 
> can't you limit the amount read / writes just by tuning the input parameters of the script (e.g.`startRun` and `maxEvents`?)

Indeed that's a nice solution, I'll give it a try as soon as I can.

cmsbuild commented:
Pull request #42909 was updated. @mandrenguyen, @consuegs, @saumyaphor4252, @jfernan2, @cmsbuild, @francescobrivio, @perrotta can you please check and sign again.

francescobrivio committed:
update BeamSpotOnlineFromOfflineConverter for multi-IOV tags

francescobrivio commented:
> > but at the cost of reading (and writing out) a lot of IOVs...would that be acceptable?
> 
> can't you limit the amount of reads / writes just by tuning the input parameters of the script (e.g.`startRun` and `maxEvents`?)

done in https://github.com/cms-sw/cmssw/pull/42909/commits/064118e709f10e968d446bb3d53662762a7e8606

francescobrivio commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42909/37066

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #42909 was updated. @mandrenguyen, @consuegs, @jfernan2, @perrotta, @francescobrivio, @saumyaphor4252 can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4efcb7/34998/summary.html
**COMMIT**: 064118e709f10e968d446bb3d53662762a7e8606
**CMSSW**: CMSSW_13_3_X_2023-10-01-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42909/34998/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - You potentially added 2 lines to the logs
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3358320
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3358295
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-01-2300+4efcb7/59091/triggerResults): no differences found

francescobrivio commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4efcb7/35008/summary.html
**COMMIT**: 064118e709f10e968d446bb3d53662762a7e8606
**CMSSW**: CMSSW_13_3_X_2023-10-02-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42909/35008/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 3 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3358320
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3358295
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-02-1100+4efcb7/59099/triggerResults): no differences found

perrotta commented:
+alca
- New unit test passes succesfully
- @francescobrivio I let you sign for db as soon as you confirm that you are fine with this

francescobrivio commented:
+db

jfernan2 commented:
+reconstruction

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
+1
